netket/sampler/autoreg.py

```diff
@@ -0,0 +1,123 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from functools import partial
+
+from netket.sampler import Sampler, SamplerState
+from netket.utils import struct
+from netket.utils.types import PRNGKeyT, PyTree
+
+import jax
+from jax import numpy as jnp
+
+
+def vmap_choice(key, a, p, replace=True):
+    """
+    p.shape: (batch, a.shape)
+    Return shape: (batch, )
+    """
+
+    def scan_fun(key, p_i):
+        new_key, key = jax.random.split(key)
+        out_i = jax.random.choice(key, a, replace=replace, p=p_i)
+        return new_key, out_i
+
+    _, out = jax.lax.scan(scan_fun, key, p)
+    return out
+
+
+@struct.dataclass
+class ARDirectSamplerState(SamplerState):
+    σ: jnp.ndarray
+    """current batch of (maybe partially sampled) configurations."""
+    cache: PyTree
+    """auxiliary states, e.g., used to implement fast autoregressive sampling."""
+    key: PRNGKeyT
+    """state of the random number generator."""
+
+    def __repr__(self):
+        return f"{type(self).__name__}(rng state={self.key})"
+
+
+@struct.dataclass
+class ARDirectSampler(Sampler):
+    """Direct sampler for autoregressive neural networks."""
+
+    def _init_cache(sampler, model, σ, key):
+        variables = model.init(key, σ)
+        if "cache" in variables:
+            _, cache = variables.pop("cache")
+        else:
+            cache = None
+        return cache
+
+    def _init_state(sampler, model, params, key):
+        new_key, key = jax.random.split(key)
+        σ = jnp.zeros((sampler.n_chains, sampler.hilbert.size), dtype=sampler.dtype)
+        cache = sampler._init_cache(model, σ, key)
+        return ARDirectSamplerState(σ=σ, cache=cache, key=new_key)
+
+    def _reset(sampler, model, params, state):
+        return state
+
+    def _sample_chain(sampler, model, params, state, chain_length):
+        return _sample_chain(sampler, model, params, state, chain_length)
+
+    def _sample_next(sampler, model, params, state):
+        σ, new_state = sampler._sample_chain(model, params, state, 1)
+        σ = σ.squeeze(axis=0)
+        return new_state, σ
+
+
+@partial(jax.jit, static_argnums=(1, 4))
```

Comment:
sorry I am bit confused: why this function is defined outside of the  class and called before?

Reply:
This is because if you jit the class method every time you create a new sampler (Even if identical) you have to jit it again because it's technically another method (attached to another class).

This trampoline allows us to jit only once.

It's ugly but the only way